lib/livebook_web/live/session_live.ex

```diff
@@ -1,21 +1,32 @@
 defmodule LivebookWeb.SessionLive do
   use LivebookWeb, :live_view
 
-  alias Livebook.{SessionSupervisor, Session, Delta, Notebook, Runtime}
+  import LivebookWeb.LiveHelpers
+
+  alias Livebook.{SessionSupervisor, Session, Delta, Notebook, Runtime, Users}
 
   @impl true
-  def mount(%{"id" => session_id}, _session, socket) do
+  def mount(%{"id" => session_id}, %{"current_user_id" => current_user_id}, socket) do
     if SessionSupervisor.session_exists?(session_id) do
       data =
         if connected?(socket) do
-          data = Session.register_client(session_id, self())
+          data = Session.register_client(session_id, {current_user_id, self()})
           Phoenix.PubSub.subscribe(Livebook.PubSub, "sessions:#{session_id}")
 
           data
         else
           Session.get_data(session_id)
         end
 
+      current_user = Users.fetch!(current_user_id)
+      users_map = clients_to_users_map(data.clients)
+
+      if connected?(socket) do
+        for {user_id, _user} <- users_map do
+          Phoenix.PubSub.subscribe(Livebook.PubSub, "users:#{user_id}")
```

Comment:
This will change with presence but we typically re-use the same channel (i.e. the session channel) or create a `#{session_id}_presence` channel to broadcast presences instead of doing it on the user channel. Especially because the user may leave one session but still remain in others!

Reply:
And for user updates, what most likely will happen is that you will broadcast on all sessions the user is in.

Reply:
Btw, I did more thinking about the presence stuff. You will have to use the Phoenix.Tracker because we want to store the user data on ETS. I assume the logic will go like this:

1. The tracker topic is the `users:#{user_id}`

2. In the metadata, we will have the current session the user is in (or the homepage) as well as his name, color, and a timestamp

Remember that each page the user is in will be a different entry in the tracker. Also note that updating the user metadata (such as their name and avatar color) is done with a join+leave.

With the basic setup in hand, we need to do two distinct tasks:

1. We need to track in each individual session whenever the user joins or leave it. We don't need to use tracker/presence for this, given this is already tracked by the sessions themselves, so I assume we will keep this as is

2. We need to keep the user info (name+avatar color) and send updates to all sessions, including the home page, whenever this changes. We will focus on this for now.

## Tracking user info

There are two approaches to solve this problem.

### Option 1: ETS

One option is to have a ETS table that keeps the user info. We will implement the init+handle_diff of Phoenix.Tracker. `init` will create an ETS table and `handle_diff` will update it. Now every time the user joins, you will see the timestamp of the user metadata in the new join and store it in the ETS table - as long as the new metadata timestamp is more recent than whatever is on ETS.

If you update the ETS table, you now have to send a message to all sessions that the current user is in. This can be done by calling `get_by_key(...)` and sending them a regular message. However, keep in mind that handle_diff will be called per node, so you should notify only the sessions in the current node. The other nodes will be notified by their tracker. Something like this (pseudo code!!!):

```elixir
for user <- joins do
  if user_updated?(user, ets) do
    for %{"session_id" => session_id} <- Tracker.get_by_key(user.id),
        pid = :global.whereis(session_id),
        node(pid) == node() do
      send(pid, user_updated_metadata)
    end
  end
end
```

Something else you need to keep in mind is data races. Once the user joins a session, which one do you notify first, the tracker or the session? It should be the tracker, as the information must be stored on ETS before the session tries to access said user. However, on shutdowns, there is no guarantee about the order those processes will notice the user leaving, so you should be prepared for the ETS table to be empty whenever you query user information. This takes us to option 2.

### Option 2: decentralized

As we discussed, we already know how to track users on each session. The only problem we need to solve is to detect if the user have updated their info. As I mentioned, an update is always a leave+join, and those are always in the same handle_diff.

This means you can use `handle_diff` to simply detect updates and then notify all sessions with updated metadata. Something like this (pseudo code!!!):

```elixir
leavers =
  for user <- leaves,
      %{"id" => user_id, "session_id" => session_id} <- meta,
      into: MapSet.new,
      do: {user_id, session_id}

for user <- joins,
    %{"id" => user_id, "session_id" => session_id} <- meta,
    {user_id, session_id} in leavers do
  for %{"session_id" => session_id} <- Tracker.get_by_key(user.id),
      pid = :global.whereis(session_id),
      node(pid) == node() do
    send(pid, user_updated_metadata)
  end
end
```

At this moment I am thinking it is probably simpler to not use the ETS at all.